pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -366,19 +366,36 @@ type reconstructedVolume struct {
 	blockVolumeMapper   volumepkg.BlockVolumeMapper
 }
 
+type globalVolumeInfo struct {
+	volumeName        v1.UniqueVolumeName
+	volumeSpec        *volumepkg.Spec
+	devicePath        string
+	mounter           volumepkg.Mounter
+	deviceMounter     volumepkg.DeviceMounter
+	blockVolumeMapper volumepkg.BlockVolumeMapper
+	podVolumes        map[volumetypes.UniquePodName]*reconstructedVolume
+}
+
+func (gvi *globalVolumeInfo) addPodVolume(rcv *reconstructedVolume) {
+	if gvi.podVolumes == nil {
+		gvi.podVolumes = map[volumetypes.UniquePodName]*reconstructedVolume{}
+	}
+	gvi.podVolumes[rcv.podName] = rcv
+}
+
 // syncStates scans the volume directories under the given pod directory.
 // If the volume is not in desired state of world, this function will reconstruct
 // the volume related information and put it in both the actual and desired state of worlds.
 // For some volume plugins that cannot support reconstruction, it will clean up the existing
 // mount points since the volume is no long needed (removed from desired state)
-func (rc *reconciler) syncStates() {
+func (rc *reconciler) syncStates(kubeletPodDir string) {
 	// Get volumes information by reading the pod's directory
-	podVolumes, err := getVolumesFromPodDir(rc.kubeletPodsDir)
+	podVolumes, err := getVolumesFromPodDir(kubeletPodDir)
 	if err != nil {
 		klog.ErrorS(err, "Cannot get volumes from disk, skip sync states for volume reconstruction")
 		return
 	}
-	volumesNeedUpdate := make(map[v1.UniqueVolumeName]*reconstructedVolume)
+	volumesNeedUpdate := make(map[v1.UniqueVolumeName]*globalVolumeInfo)
```

Comment:
how about instead of a map, using array of reconstructedVolumes. During updateStates, just iterate over the array.

Reply:
The reason I am using a map is because, I wanted to avoid the case where we are processing same volume from different possible sources. For example - a pod may use same volume more than once. A map ensures that - we are only processing volume associated with that pod only once.

Reply:
what are the the use case of the same pod use the same volume multiple volumes?  Also I feel if using array, we don't need to change any logic basically. Just in case we process volumes multiple times at certain cases , it should be ok, right? the state uses map, so it can correctly check and update the state.

Reply:
If we make `volumesNeedUpdate` an array then - we will be possibly processing same volumes multiple times, making multiple calls to api-server (`updateStates` function fetches node objects directly from api-sever) and other duplicate processing. 

I am not sure if logic has changed all that much from old code and using a map here seems better from correctness and may be slightly even from performance point of view. I think trade-off is worth it with very little downside.